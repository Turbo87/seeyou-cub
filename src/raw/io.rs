use crate::error::Result;
use crate::types::ByteOrder;
use std::io::Read;

/// Read i16 with specified byte order
pub fn read_i16<R: Read>(reader: &mut R, order: ByteOrder) -> Result<i16> {
    let mut buf = [0u8; 2];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => i16::from_le_bytes(buf),
        ByteOrder::BE => i16::from_be_bytes(buf),
    })
}

/// Read u16 with specified byte order
pub fn read_u16<R: Read>(reader: &mut R, order: ByteOrder) -> Result<u16> {
    let mut buf = [0u8; 2];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => u16::from_le_bytes(buf),
        ByteOrder::BE => u16::from_be_bytes(buf),
    })
}

/// Read i32 with specified byte order
pub fn read_i32<R: Read>(reader: &mut R, order: ByteOrder) -> Result<i32> {
    let mut buf = [0u8; 4];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => i32::from_le_bytes(buf),
        ByteOrder::BE => i32::from_be_bytes(buf),
    })
}

/// Read u32 with specified byte order
pub fn read_u32<R: Read>(reader: &mut R, order: ByteOrder) -> Result<u32> {
    let mut buf = [0u8; 4];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => u32::from_le_bytes(buf),
        ByteOrder::BE => u32::from_be_bytes(buf),
    })
}

/// Read u64 with specified byte order
pub fn read_u64<R: Read>(reader: &mut R, order: ByteOrder) -> Result<u64> {
    let mut buf = [0u8; 8];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => u64::from_le_bytes(buf),
        ByteOrder::BE => u64::from_be_bytes(buf),
    })
}

/// Read f32 (always little-endian per spec)
pub fn read_f32_le<R: Read>(reader: &mut R) -> Result<f32> {
    let mut buf = [0u8; 4];
    reader.read_exact(&mut buf)?;
    Ok(f32::from_le_bytes(buf))
}

/// Read u8
pub fn read_u8<R: Read>(reader: &mut R) -> Result<u8> {
    let mut buf = [0u8; 1];
    reader.read_exact(&mut buf)?;
    Ok(buf[0])
}

/// Read fixed-length byte array
pub fn read_bytes<R: Read>(reader: &mut R, len: usize) -> Result<Vec<u8>> {
    let mut buf = vec![0u8; len];
    reader.read_exact(&mut buf)?;
    Ok(buf)
}

/// Read string from bytes, attempting UTF-8 first, falling back to CP1252
pub fn read_string<R: Read>(reader: &mut R, len: usize) -> Result<String> {
    let bytes = read_bytes(reader, len)?;

    // Try UTF-8 first
    match String::from_utf8(bytes.clone()) {
        Ok(s) => Ok(s),
        Err(_) => {
            // Fallback to CP1252 (Windows-1252)
            let (decoded, _encoding, _had_errors) = encoding_rs::WINDOWS_1252.decode(&bytes);
            Ok(decoded.into_owned())
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Cursor;

    #[test]
    fn test_read_i16_le() {
        let data = vec![0x34, 0x12];
        let mut cursor = Cursor::new(data);
        assert_eq!(read_i16(&mut cursor, ByteOrder::LE).unwrap(), 0x1234);
    }

    #[test]
    fn test_read_i16_be() {
        let data = vec![0x12, 0x34];
        let mut cursor = Cursor::new(data);
        assert_eq!(read_i16(&mut cursor, ByteOrder::BE).unwrap(), 0x1234);
    }

    #[test]
    fn test_read_i32_le() {
        let data = vec![0x78, 0x56, 0x34, 0x12];
        let mut cursor = Cursor::new(data);
        assert_eq!(read_i32(&mut cursor, ByteOrder::LE).unwrap(), 0x12345678);
    }

    #[test]
    fn test_read_f32_le() {
        let value = std::f32::consts::PI;
        let bytes = value.to_le_bytes();
        let mut cursor = Cursor::new(bytes);
        let result = read_f32_le(&mut cursor).unwrap();
        assert!((result - value).abs() < 0.0001);
    }

    #[test]
    fn test_read_string_utf8() {
        let data = b"Hello";
        let mut cursor = Cursor::new(data);
        assert_eq!(read_string(&mut cursor, 5).unwrap(), "Hello");
    }

    #[test]
    fn test_read_string_cp1252_fallback() {
        // CP1252 character (not valid UTF-8)
        let data = vec![0xE9]; // é in CP1252
        let mut cursor = Cursor::new(data);
        assert_eq!(read_string(&mut cursor, 1).unwrap(), "é");
    }
}
