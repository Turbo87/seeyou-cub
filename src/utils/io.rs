use crate::types::ByteOrder;
use std::io::{Read, Result, Write};

/// Read i16 with specified byte order
pub fn read_i16<R: Read>(reader: &mut R, order: ByteOrder) -> Result<i16> {
    let mut buf = [0u8; 2];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => i16::from_le_bytes(buf),
        ByteOrder::BE => i16::from_be_bytes(buf),
    })
}

/// Read u16 with specified byte order
pub fn read_u16<R: Read>(reader: &mut R, order: ByteOrder) -> Result<u16> {
    let mut buf = [0u8; 2];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => u16::from_le_bytes(buf),
        ByteOrder::BE => u16::from_be_bytes(buf),
    })
}

/// Read i32 with specified byte order
pub fn read_i32<R: Read>(reader: &mut R, order: ByteOrder) -> Result<i32> {
    let mut buf = [0u8; 4];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => i32::from_le_bytes(buf),
        ByteOrder::BE => i32::from_be_bytes(buf),
    })
}

/// Read u32 with specified byte order
pub fn read_u32<R: Read>(reader: &mut R, order: ByteOrder) -> Result<u32> {
    let mut buf = [0u8; 4];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => u32::from_le_bytes(buf),
        ByteOrder::BE => u32::from_be_bytes(buf),
    })
}

/// Read u64 with specified byte order
pub fn read_u64<R: Read>(reader: &mut R, order: ByteOrder) -> Result<u64> {
    let mut buf = [0u8; 8];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => u64::from_le_bytes(buf),
        ByteOrder::BE => u64::from_be_bytes(buf),
    })
}

/// Read f32 (always little-endian per spec)
pub fn read_f32_le<R: Read>(reader: &mut R) -> Result<f32> {
    let mut buf = [0u8; 4];
    reader.read_exact(&mut buf)?;
    Ok(f32::from_le_bytes(buf))
}

/// Read u8
pub fn read_u8<R: Read>(reader: &mut R) -> Result<u8> {
    let mut buf = [0u8; 1];
    reader.read_exact(&mut buf)?;
    Ok(buf[0])
}

/// Write u8
pub fn write_u8<W: Write>(writer: &mut W, value: u8) -> Result<()> {
    writer.write_all(&[value])
}

/// Write i16 with specified byte order
pub fn write_i16<W: Write>(writer: &mut W, value: i16, order: ByteOrder) -> Result<()> {
    let buf = match order {
        ByteOrder::LE => value.to_le_bytes(),
        ByteOrder::BE => value.to_be_bytes(),
    };
    writer.write_all(&buf)
}

/// Write u16 with specified byte order
pub fn write_u16<W: Write>(writer: &mut W, value: u16, order: ByteOrder) -> Result<()> {
    let buf = match order {
        ByteOrder::LE => value.to_le_bytes(),
        ByteOrder::BE => value.to_be_bytes(),
    };
    writer.write_all(&buf)
}

/// Write i32 with specified byte order
pub fn write_i32<W: Write>(writer: &mut W, value: i32, order: ByteOrder) -> Result<()> {
    let buf = match order {
        ByteOrder::LE => value.to_le_bytes(),
        ByteOrder::BE => value.to_be_bytes(),
    };
    writer.write_all(&buf)
}

/// Write u32 with specified byte order
pub fn write_u32<W: Write>(writer: &mut W, value: u32, order: ByteOrder) -> Result<()> {
    let buf = match order {
        ByteOrder::LE => value.to_le_bytes(),
        ByteOrder::BE => value.to_be_bytes(),
    };
    writer.write_all(&buf)
}

/// Write u64 with specified byte order
pub fn write_u64<W: Write>(writer: &mut W, value: u64, order: ByteOrder) -> Result<()> {
    let buf = match order {
        ByteOrder::LE => value.to_le_bytes(),
        ByteOrder::BE => value.to_be_bytes(),
    };
    writer.write_all(&buf)
}

/// Write f32 (always little-endian per spec)
pub fn write_f32_le<W: Write>(writer: &mut W, value: f32) -> Result<()> {
    writer.write_all(&value.to_le_bytes())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Cursor;

    #[test]
    fn test_read_i16_le() {
        let data = vec![0x34, 0x12];
        let mut cursor = Cursor::new(data);
        assert_eq!(read_i16(&mut cursor, ByteOrder::LE).unwrap(), 0x1234);
    }

    #[test]
    fn test_read_i16_be() {
        let data = vec![0x12, 0x34];
        let mut cursor = Cursor::new(data);
        assert_eq!(read_i16(&mut cursor, ByteOrder::BE).unwrap(), 0x1234);
    }

    #[test]
    fn test_read_i32_le() {
        let data = vec![0x78, 0x56, 0x34, 0x12];
        let mut cursor = Cursor::new(data);
        assert_eq!(read_i32(&mut cursor, ByteOrder::LE).unwrap(), 0x12345678);
    }

    #[test]
    fn test_read_f32_le() {
        let value = std::f32::consts::PI;
        let bytes = value.to_le_bytes();
        let mut cursor = Cursor::new(bytes);
        let result = read_f32_le(&mut cursor).unwrap();
        assert!((result - value).abs() < 0.0001);
    }

    #[test]
    fn test_write_u8() {
        let mut buf = Vec::new();
        write_u8(&mut buf, 0x42).unwrap();
        assert_eq!(buf, vec![0x42]);
    }

    #[test]
    fn test_write_i16_le() {
        let mut buf = Vec::new();
        write_i16(&mut buf, 0x1234, ByteOrder::LE).unwrap();
        assert_eq!(buf, vec![0x34, 0x12]);
    }

    #[test]
    fn test_write_i16_be() {
        let mut buf = Vec::new();
        write_i16(&mut buf, 0x1234, ByteOrder::BE).unwrap();
        assert_eq!(buf, vec![0x12, 0x34]);
    }

    #[test]
    fn test_write_u16_le() {
        let mut buf = Vec::new();
        write_u16(&mut buf, 0x1234, ByteOrder::LE).unwrap();
        assert_eq!(buf, vec![0x34, 0x12]);
    }

    #[test]
    fn test_write_u16_be() {
        let mut buf = Vec::new();
        write_u16(&mut buf, 0x1234, ByteOrder::BE).unwrap();
        assert_eq!(buf, vec![0x12, 0x34]);
    }

    #[test]
    fn test_write_i32_le() {
        let mut buf = Vec::new();
        write_i32(&mut buf, 0x12345678, ByteOrder::LE).unwrap();
        assert_eq!(buf, vec![0x78, 0x56, 0x34, 0x12]);
    }

    #[test]
    fn test_write_i32_be() {
        let mut buf = Vec::new();
        write_i32(&mut buf, 0x12345678, ByteOrder::BE).unwrap();
        assert_eq!(buf, vec![0x12, 0x34, 0x56, 0x78]);
    }

    #[test]
    fn test_write_u32_le() {
        let mut buf = Vec::new();
        write_u32(&mut buf, 0x12345678, ByteOrder::LE).unwrap();
        assert_eq!(buf, vec![0x78, 0x56, 0x34, 0x12]);
    }

    #[test]
    fn test_write_u32_be() {
        let mut buf = Vec::new();
        write_u32(&mut buf, 0x12345678, ByteOrder::BE).unwrap();
        assert_eq!(buf, vec![0x12, 0x34, 0x56, 0x78]);
    }

    #[test]
    fn test_write_u64_le() {
        let mut buf = Vec::new();
        write_u64(&mut buf, 0x123456789ABCDEF0, ByteOrder::LE).unwrap();
        assert_eq!(buf, vec![0xF0, 0xDE, 0xBC, 0x9A, 0x78, 0x56, 0x34, 0x12]);
    }

    #[test]
    fn test_write_u64_be() {
        let mut buf = Vec::new();
        write_u64(&mut buf, 0x123456789ABCDEF0, ByteOrder::BE).unwrap();
        assert_eq!(buf, vec![0x12, 0x34, 0x56, 0x78, 0x9A, 0xBC, 0xDE, 0xF0]);
    }

    #[test]
    fn test_write_f32_le() {
        let value = std::f32::consts::PI;
        let mut buf = Vec::new();
        write_f32_le(&mut buf, value).unwrap();
        assert_eq!(buf, value.to_le_bytes());
    }
}
