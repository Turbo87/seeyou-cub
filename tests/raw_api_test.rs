use insta::assert_debug_snapshot;
use seeyou_cub::{Header, Item, RawItemData};
use std::collections::HashMap;
use std::fs::File;
use std::io::{Seek, SeekFrom};

#[test]
fn raw_api_read_all_items() {
    let mut file =
        File::open("tests/fixtures/france_2024.07.02.cub").expect("Failed to open fixture");

    let header = Header::read(&mut file).unwrap();

    // Read all items to verify we can parse the entire file
    let mut count = 0;
    for i in 0..header.hdr_items {
        let offset = header.header_offset as u64 + (i as u64 * header.size_of_item as u64);
        file.seek(SeekFrom::Start(offset)).unwrap();

        let item =
            Item::read(&mut file, &header).unwrap_or_else(|_| panic!("Failed to read item {}", i));

        let offset = header.data_offset as u64 + item.points_offset as u64;
        file.seek(SeekFrom::Start(offset)).unwrap();

        let _item_data = RawItemData::read(&mut file, &header)
            .unwrap_or_else(|_| panic!("Failed to read item data for item {}", i));

        count += 1;
    }

    assert_eq!(count, 1368);
}

#[test]
fn parse_france_fixture_raw_api() {
    let mut file =
        File::open("tests/fixtures/france_2024.07.02.cub").expect("Failed to open fixture");

    let header = Header::read(&mut file).expect("Failed to read header");

    assert_debug_snapshot!("raw_header", header);

    // Read all items
    let mut items = Vec::new();
    for i in 0..header.hdr_items {
        let offset = header.header_offset as u64 + (i as u64 * header.size_of_item as u64);
        file.seek(SeekFrom::Start(offset)).unwrap();

        let item =
            Item::read(&mut file, &header).unwrap_or_else(|_| panic!("Failed to read item {}", i));

        items.push(item);
    }

    assert_eq!(items.len(), 1368);
    assert_debug_snapshot!("raw_items_first_5", &items[0..5]);
    assert_debug_snapshot!("raw_items_last_5", &items[items.len() - 5..]);

    // Airspace type distribution
    let mut class_counts = HashMap::new();
    for item in &items {
        *class_counts.entry(item.class()).or_insert(0) += 1;
    }
    let mut class_counts = class_counts.iter().collect::<Vec<_>>();
    class_counts.sort_by(|a, b| a.1.cmp(b.1).reverse());
    assert_debug_snapshot!("raw_class_counts", class_counts);

    let mut style_counts = HashMap::new();
    for item in &items {
        *style_counts.entry(item.style()).or_insert(0) += 1;
    }
    let mut style_counts = style_counts.iter().collect::<Vec<_>>();
    style_counts.sort_by(|a, b| a.1.cmp(b.1).reverse());
    assert_debug_snapshot!("raw_style_counts", style_counts);

    // First item complete data
    let first_item = items.first().unwrap();
    let offset = header.data_offset as u64 + first_item.points_offset as u64;
    file.seek(SeekFrom::Start(offset)).unwrap();
    let first_item_data =
        RawItemData::read(&mut file, &header).expect("Failed to read first item data");
    assert_debug_snapshot!("raw_first_item_data", first_item_data);

    // Last item complete data
    let last_item = items.last().unwrap();
    let offset = header.data_offset as u64 + last_item.points_offset as u64;
    file.seek(SeekFrom::Start(offset)).unwrap();
    let last_item_data =
        RawItemData::read(&mut file, &header).expect("Failed to read last item data");
    assert_debug_snapshot!("raw_last_item_data", last_item_data);

    // Collect all names and auto-select representatives
    let mut all_names = Vec::new();
    let mut min_points_item: (usize, usize) = (0, usize::MAX); // (index, point_ops_count)
    let mut max_points_item: (usize, usize) = (0, 0);

    for (i, item) in items.iter().enumerate() {
        let offset = header.data_offset as u64 + item.points_offset as u64;
        file.seek(SeekFrom::Start(offset)).unwrap();

        let item_data = RawItemData::read(&mut file, &header)
            .unwrap_or_else(|_| panic!("Failed to read item data for item {}", i));

        // Decode name from raw bytes
        let name = item_data
            .name
            .as_ref()
            .map(|bytes| String::from_utf8_lossy(bytes).to_string())
            .unwrap_or_default();
        all_names.push(name);

        // Track min/max by point_ops count
        let point_ops_count = item_data.point_ops.len();
        if min_points_item.1 > point_ops_count {
            min_points_item = (i, point_ops_count);
        }
        if max_points_item.1 < point_ops_count {
            max_points_item = (i, point_ops_count);
        }
    }

    all_names.sort();
    assert_debug_snapshot!("raw_all_airspace_names", all_names);

    // Snapshot smallest airspace by point_ops count
    let item = &items[min_points_item.0];
    let offset = header.data_offset as u64 + item.points_offset as u64;
    file.seek(SeekFrom::Start(offset)).unwrap();
    let item_data =
        RawItemData::read(&mut file, &header).expect("Failed to read smallest item data");
    assert_debug_snapshot!("raw_representative_smallest_by_points", item_data);

    // Snapshot largest airspace by point_ops count
    let item = &items[max_points_item.0];
    let offset = header.data_offset as u64 + item.points_offset as u64;
    file.seek(SeekFrom::Start(offset)).unwrap();
    let item_data =
        RawItemData::read(&mut file, &header).expect("Failed to read largest item data");
    assert_debug_snapshot!("raw_representative_largest_by_points", item_data);
}
