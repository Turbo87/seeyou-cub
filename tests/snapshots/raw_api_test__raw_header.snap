---
source: tests/raw_api_test.rs
expression: header
---
Header {
    title: b"SeeYou Airspace \xa9 Naviter 2011",
    allowed_serials: [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
    ],
    pc_byte_order: 1,
    crc32: 0,
    key: [
        142,
        73,
        51,
        123,
        14,
        15,
        61,
        172,
        100,
        127,
        208,
        179,
        93,
        0,
        197,
        191,
    ],
    size_of_item: 30,
    size_of_point: 5,
    hdr_items: 1368,
    max_pts: 1525,
    left: -0.085230246,
    top: 0.89215416,
    right: 0.17016976,
    bottom: 0.71856177,
    max_width: 0.06355881,
    max_height: 0.040743887,
    lo_la_scale: 5.3264847e-7,
    header_offset: 210,
    data_offset: 41250,
    alignment: 0,
}
